src/nanotron/generation/decode.py

```diff
@@ -243,12 +244,19 @@ def decode_text(
                     new_decoder_states.append(state)
                     # Get the new logits
                     if generation_config.use_cache:
-                        with attach_store(model=model, store=state.store):
-                            # transpose: [sequence_length, batch_size, vocab_size] -> [batch_size, sequence_length, vocab_size]
+                        if hasattr(model, "_store"):
+                            # Some model like mamba already has store build up for generation
```

Comment:
why not just keep attach_store and use it? we dont need to build it manually no?

Reply:
I need to update `store` value at `MambaModel` and `Mamba` class. Making those classes inherit `AttachableStore` doesn't share `store` across the classes. Thus, I had to create it once during `MambaForTraining` init and pass it as argument to to all the other classes

This implies the `if/else` you highlighted as in `run_generate.py` we first build the model (which create a `store` in `MambaForTraining` init) before getting into `decode_text` which create a new `store` at `with attach_store()`

Do you have any suggestion to make it cleaner ?

Reply:
okay found a cleaner way, see commit [26c1f3e](https://github.com/huggingface/nanotron/pull/103/commits/26c1f3e3e35ed1b5a2c37e013fd91d6bf989c2c3)